### PR TITLE
Use new session store

### PR DESCRIPTION
**Pull request by nwalters512**
This PR transitions us from using `express-session` to the home-grown session solution introduced in #8444.

This required introducing a new session store. I also opted to use a brand-new database table that contains more columns and that is named more in like with PL conventions. Once this is fully deployed, I'll drop the old `pl_sessions` table.

## Deployment notes

Sessions are currently used only for storing flash messages. Given that, we won't pursue any kind of sophisticated migration from the old store to the new. We're using a differently-named cookie, so there should be no overlap in data. Additionally, our usage of sticky load balancing means that pages should generally have both `POST` and `GET` requests handled by the same instance, so we shouldn't end up in a case where a server with old sessions needs to read a session written to a new session, or vice versa. Worst case, we do end up dropping a flash message, which isn't critical.

### CODE REVIEW

File: apps/prairielearn/src/migrations/20230922214925_user_sessions__create.sql [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+    updated_at TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT NOW(),
+    expires_at TIMESTAMP WITH TIME ZONE NOT NULL,
+    user_id BIGINT REFERENCES users (user_id) ON UPDATE CASCADE ON DELETE CASCADE,
+    data JSONB NOT NULL,
+    UNIQUE (session_id)
+  );
+
+CREATE INDEX IF NOT EXISTS user_sessions_user_id_updated_at_idx ON user_sessions (user_id, updated_at);
+
+CREATE INDEX IF NOT EXISTS user_sessions_updated_at_idx ON user_sessions (updated_at);
```
Comment (nwalters512) [inline apps/prairielearn/src/migrations/20230922214925_user_sessions__create.sql:15]: Expected access patterns:

- Find all recent sessions for a given user
- Find all recent sessions

I'm less sure if the latter index is required - maybe it's completely unnecessary?

File: apps/prairielearn/src/server.js [hunk 2]
```diff
@@ -150,22 +150,30 @@ module.exports.initExpress = function () {
...
         sameSite: config.sessionCookieSameSite,
       },
     }),
   );
 
+  app.use((req, res, next) => {
+    // If the session is going to expire in the near future, we'll extend it
+    // automatically for the user.
+    //
+    // TODO: make this configurable?
```
Comment (nwalters512) [inline apps/prairielearn/src/server.js:170]: 👀 should this period be configurable?

File: apps/prairielearn/src/migrations/20230922214925_user_sessions__create.sql (hunk 1 above)
Comment (mwest1066) [reply apps/prairielearn/src/migrations/20230922214925_user_sessions__create.sql:15]: Yeah, feels like its not needed to me.

File: apps/prairielearn/src/lib/authn.js [hunk 3]
```diff
@@ -65,6 +65,11 @@ module.exports.loadUser = async (req, res, authnParams, optionsParams = {}) => {
     throw new Error('user not found with user_id ' + user_id);
   }
 
+  // The session store will pick this up and store it in the `user_sessions.user_id` column.
+  if (req.session) {
+    req.session.user_id = user_id;
+  }
```
[the remark this answers is not in the document]
Comment (mwest1066) [reply apps/prairielearn/src/lib/authn.js:71]: Seems fine to me (at least until we rewrite authn to use sessions natively).

File: apps/prairielearn/src/server.js (hunk 2 above)
Comment (mwest1066) [reply apps/prairielearn/src/server.js:170]: Probably? At some point we might need to be able to set it on a per-institution basis, but I'd defer all of this for as long as possible. I'm fine hard coding it for now.

**Review by mwest1066 — APPROVED** [review]

File: apps/prairielearn/src/migrations/20230922214925_user_sessions__create.sql (hunk 1 above)
Comment (trombonekenny) [reply apps/prairielearn/src/migrations/20230922214925_user_sessions__create.sql:15]: You took the `pl_` naming out of the table name. You don't anticipate using this library with other Prairie stuff, potentially in the same database?

Comment (nwalters512) [reply apps/prairielearn/src/migrations/20230922214925_user_sessions__create.sql:15]: Sounds good, I'll remove it.

File: apps/prairielearn/src/server.js (hunk 2 above)
Comment (nwalters512) [reply apps/prairielearn/src/server.js:170]: Sounds good, I'll leave it hardcoded for now.

File: apps/prairielearn/src/migrations/20230922214925_user_sessions__create.sql (hunk 1 above)
Comment (nwalters512) [reply apps/prairielearn/src/migrations/20230922214925_user_sessions__create.sql]: Removed in f4a59fe.

**Pull request merged**